global_helpers/panther.py

```diff
@@ -160,3 +156,151 @@ def set_counter_expiration(key: str, epoch_seconds: int) -> None:
+    """
+    if isinstance(val, str):
+        item_value = {val}
+    else:
+        item_value = set(val)
+        if len(item_value) == 0:
+            # We can't remove empty sets, just return the existing value instead
+            return get_string_set(key)
+
+    response = kv_table().update_item(
+        Key={'key': key},
+        ReturnValues='UPDATED_NEW',
+        UpdateExpression='DELETE #col :ss',
+        ExpressionAttributeNames={'#col': _STRING_SET_COL},
+        ExpressionAttributeValues={':ss': item_value},
+    )
+    return response['Attributes'][_STRING_SET_COL]
+
+
+def reset_string_set(key: str) -> None:
+    """Reset a string set to empty."""
+    kv_table().update_item(
+        Key={'key': key},
+        UpdateExpression='REMOVE #col',
+        ExpressionAttributeNames={'#col': _STRING_SET_COL},
+    )
+
+
+def _test_kv_store():
+    """Integration tests which validate the functions which interact with the key-value store.
+
+    Deploy Panther and then simply run "python3 panther.py" to test.
+    """
+    assert increment_counter('panther', 1) == 1
+    assert increment_counter('labs', 3) == 3
+    assert increment_counter('panther', -2) == -1
+    assert increment_counter('panther', 0) == -1
+    assert increment_counter('panther', 11) == 10
+
+    assert get_counter('panther') == 10
+    assert get_counter('labs') == 3
+    assert get_counter('nonexistent') == 0
+
+    reset_counter('panther')
+    reset_counter('labs')
+    assert get_counter('panther') == 0
+    assert get_counter('labs') == 0
+
+    set_key_expiration('panther', int(time.time()))
+
+    # Add elements in a list, tuple, set, or as singleton strings
+    # The same key can be used to store int counts and string sets
+    assert add_to_string_set('panther', ['a', 'b']) == {'a', 'b'}
+    assert add_to_string_set('panther', ['b', 'a']) == {'a', 'b'}
+    assert add_to_string_set('panther', 'c') == {'a', 'b', 'c'}
+    assert add_to_string_set('panther', set()) == {'a', 'b', 'c'}
+    assert add_to_string_set('panther', {'b', 'c', 'd'}) == {'a', 'b', 'c', 'd'}
+    assert add_to_string_set('panther', ('d', 'e')) == {'a', 'b', 'c', 'd', 'e'}
+
+    # Empty strings are allowed
+    assert add_to_string_set('panther', '') == {'a', 'b', 'c', 'd', 'e', ''}
+
+    assert get_string_set('labs') == set()
+    assert get_string_set('panther') == {'a', 'b', 'c', 'd', 'e', ''}
+
+    assert remove_from_string_set('panther', ['b', 'c', 'd']) == {'a', 'e', ''}
+    assert remove_from_string_set('panther', '') == {'a', 'e'}
+    assert remove_from_string_set('panther', '') == {'a', 'e'}
+
+    # Overwrite contents completely
+    put_string_set('panther', ['go', 'python'])
+    assert get_string_set('panther') == {'go', 'python'}
+    put_string_set('labs', [])
+    assert get_string_set('labs') == set()
+
+    reset_string_set('panther')
+    reset_string_set('nonexistent')  # no error
+    assert get_string_set('panther') == set()
+
+
+if __name__ == '__main__':
```

Comment:
Can we include this and the tests in a separate file? IIRC this file is shown in the UI cc @nhakmiller

Reply:
After my most recent change this is no longer shown in the UI, it is saved as `panther_oss_helpers` and only accessible from the CLI.

Reply:
I'm happy to move this somewhere else in the future, I just wanted to put it somewhere